src/ai/backend/manager/models/gql_models/container_registry_utils.py

```diff
@@ -0,0 +1,255 @@
+    project: str
+    project_id: uuid.UUID
+
+    def __init__(self, db_sess: SASession, scope_id: ScopeType):
+        if not isinstance(scope_id, ProjectScope):
+            raise NotImplementedAPI("Quota mutation currently supports only the project scope.")
+
+        self.db_sess = db_sess
+        self.scope_id = scope_id
+
+    async def __ainit__(self) -> None:
+        from ...container_registry import ContainerRegistryRow
+
+        assert isinstance(self.scope_id, ProjectScope)
+
+        project_id = self.scope_id.project_id
+        group_query = (
+            sa.select(GroupRow)
+            .where(GroupRow.id == project_id)
+            .options(load_only(GroupRow.container_registry))
+        )
+        result = await self.db_sess.execute(group_query)
+        group_row = result.scalar_one_or_none()
+
+        if not HarborQuotaManager._is_valid_group_row(group_row):
+            raise ContainerRegistryNotFound(
+                f"Container registry info does not exist or is invalid in the group. (gr: {project_id})"
+            )
+
+        registry_name, project = (
+            group_row.container_registry["registry"],
+            group_row.container_registry["project"],
+        )
+
+        registry_query = sa.select(ContainerRegistryRow).where(
+            (ContainerRegistryRow.registry_name == registry_name)
+            & (ContainerRegistryRow.project == project)
+        )
+
+        result = await self.db_sess.execute(registry_query)
+        registry = result.scalars().one_or_none()
+
+        if not registry:
+            raise ContainerRegistryNotFound(
+                f"Specified container registry row not found. (cr: {registry_name}, gr: {project})"
+            )
+
+        self.group_row = group_row
+        self.registry = registry
+        self.project = project
+        self.project_id = project_id
+
+    @classmethod
+    def _is_valid_group_row(cls, group_row: GroupRow) -> bool:
+        return (
+            group_row
+            and group_row.container_registry
+            and "registry" in group_row.container_registry
+            and "project" in group_row.container_registry
+        )
+
+    async def _get_harbor_project_id(
+        self, sess: aiohttp.ClientSession, rqst_args: dict[str, Any]
+    ) -> str:
+        get_project_id_api = (
+            yarl.URL(self.registry.url) / "api" / "v2.0" / "projects" / self.project
+        )
+
+        async with sess.get(get_project_id_api, allow_redirects=False, **rqst_args) as resp:
+            if resp.status != 200:
+                raise InternalServerError(f"Failed to get harbor project_id! response: {resp}")
+
+            res = await resp.json()
+            harbor_project_id = res["project_id"]
+            return harbor_project_id
+
+    async def _get_quota_info(
+        self, sess: aiohttp.ClientSession, rqst_args: dict[str, Any]
+    ) -> HarborQuotaInfo:
+        harbor_project_id = await self._get_harbor_project_id(sess, rqst_args)
```

Comment:
Is there a better way to pass values instead of using a dictionary?
It’s unclear what values are being passed.
At the very least, the function could take a `dataclass` and convert it to a dictionary only when making an external request.